tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx

```diff
@@ -40,16 +40,22 @@
 namespace ROOT {
 namespace Experimental {
 
-class RFieldDescriptorBuilder;
 class RNTupleDescriptor;
-class RNTupleDescriptorBuilder;
 class RNTupleModel;
 
 namespace Detail {
 class RColumnElementBase;
 class RFieldBase;
-}
+} // namespace Detail
 
+namespace Internal {
+class RColumnDescriptorBuilder;
+class RColumnGroupDescriptorBuilder;
+class RClusterDescriptorBuilder;
+class RClusterGroupDescriptorBuilder;
+class RFieldDescriptorBuilder;
+class RNTupleDescriptorBuilder;
+} // namespace Internal
```

Comment:
For my own curiosity: How did this work before, ie without forward-declaring `RColumnDescriptorBuilder`?

Reply:
I'm actually not sure. After having moved the classes to `Internal`, the friend class declarations stopped working.  I.e. `friend RColumnDescriptorBuilder;` works without forward declaring the class, but `friend Internal::RColumnDescriptorBuilder` does not.

Reply:
Ah, maybe I understand after looking through https://en.cppreference.com/w/cpp/language/friend: `friend class RColumnDescriptorBuilder` should be case 3) `elaborated-class-specifier`. In this case "The name of the class that is used in this friend declaration does not need to be previously declared." On the other hand, with a namespace we get into case 4) which "will not forward declare a new type" and the compiler chooses to report an error.